README.md
[![Build Status](https://travis-ci.org/hirobanex/p5-Email-Forward-Dispatch.png?branch=master)](https://travis-ci.org/hirobanex/p5-Email-Forward-Dispatch) [![Coverage Status](https://coveralls.io/repos/hirobanex/p5-Email-Forward-Dispatch/badge.png?branch=master)](https://coveralls.io/r/hirobanex/p5-Email-Forward-Dispatch?branch=master)
# NAME

Email::Forward::Dispatch - use ~/.forward plaggerable

# SYNOPSIS

    # in /home/hirobanex/script.pl
    use Email::Forward::Dispatch;

    my $dispatcher = Email::Forward::Dispatch->new(
        is_forward_cb   => sub { ($_[1]->header('To') =~ /hirobanex\@gmail\.com/) ? 1 : 0 },
        forward_cb      => sub { print $_[1]->header('To') },
    );

    or 
    my $dispatcher = Email::Forward::Dispatch->new(
        mail      => scalar do {local $/; <STDIN>; },
        hooks_dir => "MyMailNotify::Hooks",
    );

    $dispatcher->run;



    #in /home/hirobanex/.forward
    "|exec /home/hirobanex/script.pl"

# DESCRIPTION

Email::Forward::Dispatch is Email forward utility tool. 

# LICENSE

Copyright (C) Hiroyuki Akabane.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Hiroyuki Akabane <hirobanex@gmail.com>


Changes
Revision history for Perl extension Email::Forward::Dispatch

{{$NEXT}}

0.04 2015-04-20T11:51:32Z
    - upload to cpan!!

0.03 2015-04-20T11:08:12Z
    - fixed innter method
    - change Mremora to Email::Forward::Dispatch

0.02 2013-11-19T12:03:48Z
    - version 2

0.01 2013-11-19T12:03:08Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Email-Forward-Dispatch',
    module_name     => 'Email::Forward::Dispatch',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Mremora.pm
package Mremora;
use 5.008005;
use strict;
use warnings;
use Email::MIME;
use Module::Pluggable::Object;

our $VERSION = "0.02";

my $id = 0;
sub new {
    my ($class,%args) = @_;

    my $default_hook = "";
    unless ($args{hooks_dir}) {
        $args{is_forward_cb}   or die 'you must register is_forward_cb!';
        $args{forward_cb} or die 'you must register forward_cb!';

        {
            $default_hook = "Mremora::Hooks::Default".$id++;
            no strict 'refs'; ## no critic.
            push @{"$default_hook\::ISA"}, 'Mremora::Hooks';
            *{"$default_hook\::is_forward"}   = sub { my ($class, $parsed) = @_;  $args{is_forward_cb}->($default_hook,$parsed); };
            *{"$default_hook\::forward"} = sub { my ($class, $parsed) = @_;  $args{forward_cb}->($default_hook,$parsed); };
        }
    }

    my $mail = $args{mail} || do {local $/; <STDIN>; } || die 'you must set mail option or STDIN !';

    my $self = bless +{
        email         => Email::MIME->new($mail),
        hooks_dir     => $args{hooks_dir} || 'Mremora::Hooks',
        default_hook => $default_hook,   
    }, $class;
}

sub default_hook { $_[0]->{default_hook} }

sub run {
    my ($self) = @_;

    my @hooks = $self->fetch_hooks();

    if (my $default_hook = $self->{default_hook}) {
        $default_hook->is_forward($self->{email})
            and $default_hook->forward($self->{email});
    }else{
        for my $hook (@hooks) {
            next unless $hook->is_forward($self->{email});

            $hook->forward($self->{email});
        }
    }
}

sub fetch_hooks {
    my $self = shift;

    my @hooks = Module::Pluggable::Object->new(
        require     => 1,
        search_path => $self->{hooks_dir},
    )->plugins;
}

1;
__END__

=encoding utf-8

=head1 NAME

Mremora - use ~/.forward plaggerable

=head1 SYNOPSIS

    # in /home/hirobanex/script.pl
    use Mremora;

    my $dispatcher = Mremora->new(
        is_forward_cb   => sub { ($_[1]->header('To') =~ /hirobanex\@gmail\.com/) ? 1 : 0 },
        forward_cb      => sub { print $_[1]->header('To') },
    );

    or 

    my $dispatcher = Mremora->new(
        mail      => scalar do {local $/; <STDIN>; },
        hooks_dir => "MyMailNotify::Hooks",
    );

    $dispatcher->run;


    #in /home/hirobanex/.forward
    "|exec /home/hirobanex/script.pl"

=head1 DESCRIPTION

Mremora is Email forward utility tool. 

=head1 LICENSE

Copyright (C) Hiroyuki Akabane.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Hiroyuki Akabane E<lt>hirobanex@gmail.comE<gt>

=for stopwords plaggerable

=cut



lib/Email/Forward/Dispatch.pm
package Email::Forward::Dispatch;
use 5.008005;
use strict;
use warnings;
use Email::MIME;
use Module::Pluggable::Object;

our $VERSION = "0.03";

my $id = 0;
sub new {
    my ($class,%args) = @_;

    my $default_hook = "";
    unless ($args{hooks_dir}) {
        $args{is_forward_cb}   or die 'you must register is_forward_cb!';
        $args{forward_cb} or die 'you must register forward_cb!';

        {
            $default_hook = "Email::Forward::Dispatch::Hooks::Default".$id++;
            no strict 'refs'; ## no critic.
            push @{"$default_hook\::ISA"}, 'Email::Forward::Dispatch::Hooks';
            *{"$default_hook\::is_forward"}   = sub { my ($class, $parsed) = @_;  $args{is_forward_cb}->($default_hook,$parsed); };
            *{"$default_hook\::forward"} = sub { my ($class, $parsed) = @_;  $args{forward_cb}->($default_hook,$parsed); };
        }
    }

    my $mail = $args{mail} || do {local $/; <STDIN>; } || die 'you must set mail option or STDIN !';

    my $self = bless +{
        email         => Email::MIME->new($mail),
        hooks_dir     => $args{hooks_dir} || 'Email::Forward::Dispatch::Hooks',
        default_hook => $default_hook,   
    }, $class;
}

sub default_hook { $_[0]->{default_hook} }

sub run {
    my ($self) = @_;

    my @hooks = $self->fetch_hooks();

    if (my $default_hook = $self->{default_hook}) {
        $default_hook->is_forward($self->{email})
            and $default_hook->forward($self->{email});
    }else{
        for my $hook (@hooks) {
            next unless $hook->is_forward($self->{email});

            $hook->forward($self->{email});
        }
    }
}

sub fetch_hooks {
    my $self = shift;

    my @hooks = Module::Pluggable::Object->new(
        require     => 1,
        search_path => $self->{hooks_dir},
    )->plugins;
}

1;
__END__

=encoding utf-8

=head1 NAME

Email::Forward::Dispatch - use ~/.forward plaggerable

=head1 SYNOPSIS

    # in /home/hirobanex/script.pl
    use Email::Forward::Dispatch;

    my $dispatcher = Email::Forward::Dispatch->new(
        is_forward_cb   => sub { ($_[1]->header('To') =~ /hirobanex\@gmail\.com/) ? 1 : 0 },
        forward_cb      => sub { print $_[1]->header('To') },
    );

    or 

    my $dispatcher = Email::Forward::Dispatch->new(
        mail      => scalar do {local $/; <STDIN>; },
        hooks_dir => "MyMailNotify::Hooks",
    );

    $dispatcher->run;


    #in /home/hirobanex/.forward
    "|exec /home/hirobanex/script.pl"

=head1 DESCRIPTION

Email::Forward::Dispatch is Email forward utility tool. 

=head1 LICENSE

Copyright (C) Hiroyuki Akabane.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Hiroyuki Akabane E<lt>hirobanex@gmail.comE<gt>

=for stopwords plaggerable

=cut

